### PR TITLE
Cache smart wallet account factory address

### PR DESCRIPTION
**Pull request by jakeloo**
## Changes

- Cache smart wallet account factory address (create-account)

---

## PR-Codex overview
The focus of this PR is to introduce caching mechanisms using Redis for factory addresses and deployed account codes to handle potential race conditions in account creation and transaction processing.

### Detailed summary
- Added Redis caching for factory addresses and deployed account codes
- Improved handling of potential race conditions in account creation and transaction processing

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/server/utils/wallets/getSmartWallet.ts [hunk 1]
```diff
@@ -14,13 +15,28 @@ export const getSmartWallet = async ({
...
+
   try {
-    const contract = await getContract({
-      chainId,
-      contractAddress: accountAddress,
-    });
-    factoryAddress = await contract.call("factory");
+    // Note: This is a temporary solution to use cached deployed address's factory address from create-account
+    // This is needed due to a potential race condition of submitting a transaction immediately after creating an account that is not yet mined onchain
+    if (redis) {
```
Comment (arcoraven) [inline src/server/utils/wallets/getSmartWallet.ts:22]: `redis` is required to run Engine now so this should never be `null` after Engine server/worker is started.

File: src/server/utils/wallets/getSmartWallet.ts [hunk 2]
```diff
@@ -14,13 +15,28 @@ export const getSmartWallet = async ({
...
+      const contract = await getContract({
+        chainId,
+        contractAddress: accountAddress,
+      });
+      factoryAddress = await contract.call("factory");
+    }
+  } catch {
+  }
+
+  if (!factoryAddress) {
```
Comment (arcoraven) [inline src/server/utils/wallets/getSmartWallet.ts:39]: This try/catch can just wrap the `throw` below like it was before:

```typescript
try
  if (!factoryAddress)
    ...
catch
  throw new Error
```

**Review by arcoraven — APPROVED** [review]
Looks correct but left minor polish comments

**Review by arcoraven — APPROVED** [review]

**Review by farhanW3 — APPROVED** [review]
lgtm

**Pull request merged**